src/base.jl

```diff
@@ -291,15 +339,16 @@ function IS.add_forecast!(
 
     if component isa LoadZones
         uuids = Set([IS.get_uuid(x) for x in get_buses(component)])
-        for component_ in (load for load in IS.get_components(ElectricLoad, data)
-                          if get_bus(load) |> IS.get_uuid in uuids)
+        for _component in (load for load in IS.get_components(ElectricLoad, data)
```

Comment:
What's wrong with component_?

Reply:
I have a preference for using `_component`, with underscore as a prefix, because to me, having a `_` in `_local_variable` provides some context about the scope of the `_local_variable` variable.

I have been thinking about this a little more and my thoughts are as follows. I think there's at least three places one can use `_` as a prefix to a symbol

1) Variable names
2) Field names
3) Function names

And each of these has slightly different meanings to someone trying to understand what the code is doing.

___

1) Variable names

Take the following trivial function as an example.


```julia
"""
Calculate total of elements that are less than or equal to 5
"""
function calculate_total(elements::Vector{Int})
    ignore = Int[]

    for (i, _element) in enumerate(elements)
        if _element > 5
            push!(ignore, i)
        end
    end

    total = 0

    for (i, _element) in enumerate(elements)
        if !(i in ignore)
            total += _element
        end
    end

   return total
end
```

There's two blocks of code here, a) the block that creates the ignore list and b) the block that sums the code.

Using the underscore in the `_element` variable clarifies that changing the variable name `_element` in block a) should not the use of `_element` in block b). It also implies that `_element` is of local scope to the `for` loop, since `_element` is not defined outside the body of the `for` loop.

```
julia> function calculate_total(elements::Vector{Int})
           ignore = Int[]

           for (i, _element) in enumerate(elements)
               if _element > 5
                   push!(ignore, i)
               end
           end


           @show _element # <- add this line to throw an error.


           total = 0

           for (i, _element) in enumerate(elements)
               if !(i in ignore)
                   total += _element
               end
           end

          return total
       end
calculate_total (generic function with 1 method)

julia> calculate_total([1, 2, 3, 4, 5, 6, 7, 8])
ERROR: UndefVarError: _element not defined
Stacktrace:
 [1] macro expansion at ./show.jl:562 [inlined]
 [2] calculate_total(::Array{Int64,1}) at ./REPL[18]:11
 [3] top-level scope at REPL[19]:1

julia>
```

This is not to say that all `for` loop element variable names should be prefixed with an underscore since that is just visual noise and is unnecessary. In a larger function body, if the terminology for a `element` needs to be repeated and used in separate contexts, using `_element` helps to clarify context and scope and can improve readability.

2) Field names

In various Object Oriented Programming (OOP) based languages, there are ways within the language to enforce visibility of fields that belong to a class. Dynamic languages (such as Julia, Python) tend to not support restricting of direct access to some of an object's fields. In this case, the convention for using an underscore is indicate that the field in question has some implementation details of the struct and off limits to an user as an interface. However, using this convention does not mean that field names without an underscore are all part of the public interface i.e. one should refer to the documentation to find what is part of the interface and what are fields are guaranteed to be stable.

Let's say we are implementing a `Vector3D` struct in Julia.

![](https://i.stack.imgur.com/RFKBk.gif)

As an example, take the following code.

````julia
"""
Define Vector3D struct

Public API:

    v = Vector3D(x, y, z)
    println(v.x)
    println(v.y)
    println(v.z)

    println(get_magnitude(v))
    println(get_polar_angle(v))
    println(get_azimuth_angle(v))
"""

export Vector3D

"""
    Vector3D(x::Int, y::Int, z::Int)

# Example

```julia
v = Vector3D(1, 1, 1)
```
"""
struct Vector3D
    x::Int
    y::Int
    z::Int
    r::Float64       # private variable
    polar::Float64   # private variable
    azimuth::Float64 # private variable
    _r_xy::Float64   # implementation details
    _r_yz::Float64   # implementation details
    _r_zx::Float64   # implementation details

    function Vector3D(x, y, z)
        r = √(x^2 + y^2 + z^2)
        return new(
            x,
            y,
            z,
            r,
            arctan(y/x),
            arccos(z/r),
            √(x^2 + y^2),
            √(y^2 + z^2),
            √(z^2 + x^2),
        )
    end
end
get_magnitude(v::Vector3D) = v.r
get_polar_angle(v::Vector3D) = v.polar
get_azimuth_angle(v::Vector3D) = v.azimuth
````

If `Vector3D` is an `export` in say `v2.1.0` and if you want to change the field names to `position_x`, `position_y` and `position_z`, then based on semantic versioning you should release a new package with version `v3.0.0`.

But based on semantic versioning, it is technically acceptable to make the following change in a patch release (i.e. `v2.1.0` -> `v2.1.1`) since your public API according to your documentation has not changed.

````julia
"""
Define Vector3D struct

Public API:

    v = Vector3D(x, y, z)
    println(v.x)
    println(v.y)
    println(v.z)

    println(get_magnitude(v))
    println(get_polar_angle(v))
    println(get_azimuth_angle(v))
"""

export Vector3D

"""
    Vector3D(x::Int, y::Int, z::Int)

# Example

```julia
v = Vector3D(1, 1, 1)
x = v.x
y = v.y
z = v.z
println(get_magnitude(v))
println(get_polar_angle(v))
println(get_azimuth_angle(v))
```
"""
struct Vector3D
    x::Int
    y::Int
    z::Int

    Vector3D(x, y, z) = new(x, y, z)
end
get_magnitude(v::Vector3D) = √(v.x^2 + v.y^2 + v.z^2)
get_polar_angle(v::Vector3D) = arctan(v.y/v.x)
get_azimuth_angle(v::Vector3D) = arccos(v.z/get_magnitude(v))
````

A user using `Vector3D` in their application / package / library should not be surprised to find that their previous code that used `v._r_x` or even `v.a_angle` broke with a new patch update, since based on the documentation the public interface defines only `x`, `y` and `z` field attributes as the API, and the remaining fields can only be accessed using the functional getter interface.

3) Function names

Finally, function names with an underscore prefix are those seem to hint that there is some abstraction that reads cleaner as a separate function but is used just a handful of times, or is a hacky solution to a problem that you want to solve temporarily in this manner but will get rid off in the future without changing anything else in the package. These underscore prefixed functions are usually local to a file in which the function is defined.

Consider the following package:

```julia
# src/Package.jl

export report_metrics
export Type1
export Type2
export Type3

include("utils.jl")
include("type1_element.jl")
include("type2_element.jl")
include("type3_element.jl")
```

With the following `src/utils.jl` file:

```julia
# src/utils.jl
abstract type Element end

function _get_int64_np_array(e::Element)
   # convert Julia type into corresponding NumPy array for metrics calculation
   ...
   return array
end

function _py_calculate_metrics(e::Element)
    pymetrics = pyimport("pymetrics")
    return pymetrics.calculate_metrics(_get_int64_np_array(e))
end

function calculate_metrics(e::Element)
    ...
    # FIXME: rewrite pymetrics in Julia for better performance
    m = _hack_metrics(e)
    return m
end
```

And the following files:

```julia
# src/type1_element.jl

struct Type1 <: Element end

function report_metrics(t1::Type1)
    ...
    m = calculate_metrics(t1)
    ...
end
```

```julia
# src/type2_element.jl

struct Type2 <: Element end

function report_metrics(t2::Type2)
    ...
    m = calculate_metrics(t2)
    ...
end
```

```julia
# src/type3_element.jl

struct Type3 <: Element end

function report_metrics(t3::Type3)
    ...
    m = calculate_metrics(t3)
    ...
end
```

- `report_metrics` is clearly part of the public API since it is exported.
- `calculate_metrics` is not part of the public API and should not be used. If a user decides to use it, they may be subjecting themselves to breaking code in the future.
- `_py_calculate_metrics` is a hacky solution to get something working temporarily
- `_get_int64_np_array` is a implementation detail required in the package that is not related to what the package does, but exists just to get the `pymetrics` package to work.

Even though `calculate_metrics` is not an `export` and may never be an `export`, it (in my opinion) should not have an underscore prefix since API is only defined by what is or is not exported / is or is not documented. Underscores as prefix to function names are relegated to implementation details that are local to an aspect of the package that is not related to the "domain" or the "application" or some hacky solution that will be removed in the future.

___

That was a brain dump of an afternoon's thought. I'm happy to discuss this further if people have more thoughts or comments regarding this.